File: Coin_Change_Brute_Force.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:

        if not coins:
            return -1
        else:
            return self.helper(coins, amount, 0, 0)

    def helper(self, coins, amount, index, count):
        if amount == 0:
            return count

        if amount < 0 or index == len(coins):
            return -1

        case1 = self.helper(coins, amount - coins[index], index, count + 1)
        case2 = self.helper(coins, amount, index + 1, count)
        if case1 == -1:
            return case2
        if case2 == -1:
            return case1
        return min(case1, case2)

File: coinchange-bruteforce.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        
        auto result = helper(coins, amount, 0, 0);
        if(result == INT_MAX) return -1;
        return result;
    }

    int helper(vector<int>& coins, int amount, int index, int coinsUsed){
        
        // base
        if(amount == 0){
            return coinsUsed;
        }
        
        if(amount<0 || index == coins.size()){
            return INT_MAX;
        }
        //case 1 take
        
        int case1 = helper(coins, amount - coins[index], index, coinsUsed + 1);
        
        //case 2 dont take
        int case2 = helper(coins, amount, index + 1, coinsUsed);
        
        return min(case1, case2);
        
        
    }
};
